unicorn/hal.c

```diff
@@ -133,40 +76,38 @@ void hal_enableIRQs (void) {
     ASSERT(sim.irqlevel);
     if( --sim.irqlevel == 0 ) {
         asm volatile ("cpsie i" : : : "memory");
-        svc(SVC_IRQ, 0, 0, 0);
+        irq();
     }
 }
 
+static uint64_t extend (uint32_t ticks) {
+    uint64_t c = timer_ticks();
+    return c + (((int32_t) ticks - (int32_t) c));
+}
+
 void hal_sleep (u1_t type, u4_t targettime) {
-    sim.xnow_cached = -1;
-    svc32(SVC_SLEEP, type, targettime, 0);
+    timer_set(extend(targettime));
+    wfi();
 }
 
 u4_t hal_ticks (void) {
     return hal_xticks();
 }
 
 u8_t hal_xticks (void) {
-    if( sim.xnow_cached < 0 ) {
-        sim.xnow_cached = svc64(SVC_TICKS, 0, 0, 0);
-    }
-    return sim.xnow_cached;
-}
-
-s2_t hal_subticks (void) {
-    return 0;
+    return timer_ticks();
 }
 
 void hal_waitUntil (u4_t time) {
     // be very strict about how long we can busy wait
     ASSERT(((s4_t) time - (s4_t) hal_ticks()) < ms2osticks(100));
-    while (1) {
+    while( 1 ) {
         u4_t now = hal_ticks();
-        if (((s4_t) (time - now)) <= 0) {
+        if( ((s4_t) (time - now)) <= 0 ) {
             return;
         }
-        sim.xnow_cached = -1;
-        svc(SVC_SLEEP, HAL_SLEEP_EXACT, time, 0);
+        // TODO - sleep
```

Comment:
This implementation does not appear to be complete. wfi() will not return until there's an interrupt, potentially much later than the intended target time.